test/test_util.h

```diff
@@ -581,6 +584,14 @@ __host__ __device__ __forceinline__ void InitValue(GenMode gen_mode, T &value, i
         }
         break;
     }
+#else
+     case RANDOM:
+     case RANDOM_BIT:
+     case RANDOM_MINUS_PLUS_ZERO:
+         _CubLog("%s\n",
+                 "cub::InitValue cannot generate random numbers on device.");
+         cub::ThreadTrap();
```

Comment:
I understand that we have to fix the build as soon as possible, so the following suggestion could be extracted into a separate issue. We could split this code into device/host supported parts instead of failing at runtime. Here is the prototype that I've been thinking about:

```cuda
enum GenModeTrivialParam
{
    UNIFORM = 0,                 // Assign to '2', regardless of integer seed
    INTEGER_SEED = 1             // Assign to integer seed
};

struct GenModeTrivial
{
    GenModeTrivialParam param;

    GenModeTrivial(GenModeTrivialParam param)
        : param(param)
    {
    }

    operator int()
    {
        return param;
    }
};

enum GenModeRandomParam
{
    RANDOM = 2,                  // Assign to random, regardless of integer seed
    RANDOM_BIT = 3,              // Assign to randomly chosen 0 or 1, regardless of integer seed
    RANDOM_MINUS_PLUS_ZERO = 4,  // Assign to random, with some values being -0.0 or +0.0 patterns
};

struct GenModeRandom
{
    GenModeRandomParam param;

    GenModeRandom(GenModeRandomParam param)
        : param(param)
    {
    }

    operator int()
    {
        return param;
    }
};

/**
 * Initialize value
 */
template <typename T>
void InitValue(GenModeRandom gen_mode, T &value, int index = 0)
{
    switch (gen_mode.param)
    {
      // RandomBits is host-only.
    case RANDOM:
        RandomBits(value);
        break;
    case RANDOM_BIT:
    {
        char c;
        RandomBits(c, 0, 0, 1);
        value = (c > 0) ? (T) 1 : (T) -1;
        break;
    }
    case RANDOM_MINUS_PLUS_ZERO:
    {
        // Replace roughly 1/128 of values with -0.0 or +0.0, and generate the rest randomly
        typedef typename cub::Traits<T>::UnsignedBits UnsignedBits;
        char c;
        RandomBits(c);
        if (c == 0)
        {
            // Replace 1/256 of values with +0.0 bit pattern
            value = SafeBitCast<T>(UnsignedBits(0));
        }
        else if (c == 1)
        {
            // Replace 1/256 of values with -0.0 bit pattern
            value = SafeBitCast<T>(UnsignedBits(UnsignedBits(1) <<
                                                (sizeof(UnsignedBits) * 8) - 1));
        }
        else
        {
            // 127/128 of values are random
            RandomBits(value);
        }
        break;
    }
    default:
         value = (T) index;
        break;
    }
}

template <typename T>
__device__ __host__ void InitValue(GenModeTrivial gen_mode, T &value, int index = 0)
{
    switch (gen_mode.param)
    {
        // RandomBits is host-only.
        case UNIFORM:
            value = 2;
            break;
        case INTEGER_SEED:
        default:
            value = (T) index;
            break;
    }
}
```

There is little to be changed in the tests. I've tested this modification on the `test/test_block_histogram.cu`, here are the changes I've made: 

```cuda
template <
    int             BINS,
    typename        SampleT,
    typename        GenMode>  // <-------------------------------------
void Initialize(

///.....

template <
    typename                    SampleT,
    int                         BINS,
    int                         BLOCK_THREADS,
    int                         ITEMS_PER_THREAD,
    BlockHistogramAlgorithm     ALGORITHM,
    typename                    GenMode> // <-------------------------------------
void Test(
```

In most cases, gen_mode is selected in place of InitValue call, so there won't be any changes at all:

```cuda
InitValue(INTEGER_SEED, h_in[i], i);
```

As an alternative, we could leave only host versions of `InitValue` (as in your previous PR) and copy an INTEGER_SEED version for `template <typename T> struct TransformOp`. It seems that it's the only place where InitValue is called from the device side. 

In both cases, we'll fail at the compilation time instead of runtime. We'll also have simpler code.

Reply:
We could also switch to an RNG implementation that works on the device. This will have to wait for a follow-up fix though.